experiment/autobumper/bumper/bumper.go

```diff
@@ -38,26 +42,201 @@ const (
+	prowRefConfigFile        = "config/prow/cluster/deck_deployment.yaml"
+	prowStagingRefConfigFile = "config/prow-staging/cluster/deck_deployment.yaml"
+)
+
+var (
+	tagRegexp = regexp.MustCompile("v[0-9]{8}-[a-f0-9]{6,9}")
 )
 
-func Call(stdout, stderr io.Writer, cmd string, args ...string) error {
+type FileArrayFlag []string
+
+func (af *FileArrayFlag) String() string {
+	return fmt.Sprint(*af)
+}
+
+func (af *FileArrayFlag) Set(value string) error {
+	for _, e := range strings.Split(value, ",") {
+		fn := strings.TrimSpace(e)
+		info, err := os.Stat(fn)
+		if err != nil {
+			return fmt.Errorf("error getting file info for %q", fn)
+		}
+		if info.IsDir() && !strings.HasSuffix(fn, string(os.PathSeparator)) {
+			fn = fn + string(os.PathSeparator)
+		}
+		*af = append(*af, fn)
+	}
+	return nil
+}
+
+type Options struct {
+	GitHubOrg     string
+	GitHubRepo    string
+	GitHubLogin   string
+	GitHubToken   string
+	GitName       string
+	GitEmail      string
+	RemoteBranch  string
+	OncallAddress string
+
+	BumpProwImages bool
+	BumpTestImages bool
+	TargetVersion  string
+
+	IncludedConfigPaths FileArrayFlag
+	ExcludedConfigPaths FileArrayFlag
+	ExtraFiles          FileArrayFlag
+
+	SkipPullRequest bool
+}
+
+func validateOptions(o *Options) error {
+	if !o.SkipPullRequest && o.GitHubToken == "" {
+		return fmt.Errorf("--github-token is mandatory when --skip-pull-request is false")
+	}
+	if !o.SkipPullRequest && (o.GitHubOrg == "" || o.GitHubRepo == "") {
+		return fmt.Errorf("--github-org and --github-repo are mandatory when --skip-pull-request is false")
+	}
+	if !o.SkipPullRequest && o.RemoteBranch == "" {
+		return fmt.Errorf("--remote-branch cannot be empty when --skip-pull-request is false")
+	}
+	if (o.GitEmail == "") != (o.GitName == "") {
+		return fmt.Errorf("--git-name and --git-email must be specified together")
+	}
+
+	if o.TargetVersion != latestVersion && o.TargetVersion != upstreamVersion &&
+		o.TargetVersion != upstreamStagingVersion && !tagRegexp.MatchString(o.TargetVersion) {
+		return fmt.Errorf("invalid --target-version, must be one of %v",
+			[]string{latestVersion, upstreamVersion, upstreamStagingVersion, tagVersion})
+	}
+	if !o.BumpProwImages && !o.BumpTestImages {
+		return fmt.Errorf("at least one of --bump-prow-images and --bump-test-images must be specified")
+	}
+	if o.BumpProwImages && o.BumpTestImages && o.TargetVersion != latestVersion {
+		return fmt.Errorf("--target-version must be latest if you want to bump both prow and test images")
+	}
+	if o.BumpTestImages && (o.TargetVersion == upstreamVersion || o.TargetVersion == upstreamStagingVersion) {
+		return fmt.Errorf("%q and %q versions can only be specified to bump prow images", upstreamVersion, upstreamStagingVersion)
+	}
+
+	if len(o.IncludedConfigPaths) == 0 {
+		return fmt.Errorf("--include-config-paths is mandatory")
+	}
+
+	return nil
+}
+
+func Run(o *Options) error {
```

Comment:
This also seems like it would be useful to unit test

Reply:
Since this function will call the `updateReferences` function as discussed in https://github.com/kubernetes/test-infra/pull/18148#discussion_r469051656, I didn't find a good way to add unit tests for it either.